src/common/service/NetRemoteDataStreamingReactors.cxx

```diff
@@ -40,4 +43,107 @@ DataStreamReader::OnCancel()
+            m_numberOfDataBlocksToStream = m_dataStreamProperties.fixedtypeproperties().numberofdatablockstostream();
+        } else {
+            HandleFailure("Invalid properties for this streaming type. Expected FixedTypeProperties for DataStreamTypeFixed");
+            return;
+        }
+
+        break;
+    }
+    case DataStreamType::DataStreamTypeContinuous: {
+        if (m_dataStreamProperties.Properties_case() == DataStreamProperties::kContinuousTypeProperties) {
+            m_numberOfDataBlocksToStream = 0;
+        } else {
+            HandleFailure("Invalid properties for this streaming type. Expected ContinuousTypeProperties for DataStreamTypeContinuous");
+            return;
+        }
+
+        break;
+    }
+    default: {
+        HandleFailure(std::format("Invalid streaming type: {}", magic_enum::enum_name(m_dataStreamProperties.type())));
+        return;
+    }
+    };
+
+    m_writeStatus.set_code(DataStreamOperationStatusCode::DataStreamOperationStatusCodeUnknown);
+    m_writeStatus.set_message("No data sent yet");
+    NextWrite();
+}
+
+void
+DataStreamWriter::OnWriteDone(bool isOk)
+{
+    // Check for a failed status code from HandleWriteFailure since that invoked a final write, thus causing this callback to be invoked.
+    if (m_writeStatus.code() == DataStreamOperationStatusCode::DataStreamOperationStatusCodeFailed) {
+        Finish(::grpc::Status::OK);
+        return;
+    }
+
+    // Continue writing if previous write was successful, otherwise handle the failure.
+    if (isOk) {
+        if (m_dataStreamProperties.type() == DataStreamType::DataStreamTypeFixed) {
+            m_numberOfDataBlocksToStream--;
+        }
+        m_writeStatus.set_code(DataStreamOperationStatusCode::DataStreamOperationStatusCodeSucceeded);
+        m_writeStatus.set_message("Data write successful");
+        NextWrite();
+    } else {
+        HandleFailure("Data write failed");
+    }
+}
+
+void
+DataStreamWriter::OnCancel()
+{
+    // The RPC is cancelled by the client, so call Finish to complete it from the server perspective.
+    Finish(grpc::Status::CANCELLED);
+}
+
+void
+DataStreamWriter::OnDone()
+{
+    delete this;
+}
+
+void
+DataStreamWriter::NextWrite()
+{
+    if (m_dataStreamProperties.type() == DataStreamType::DataStreamTypeContinuous ||
+        (m_dataStreamProperties.type() == DataStreamType::DataStreamTypeFixed && m_numberOfDataBlocksToStream > 0)) {
+        // Create data to write to the client.
+        const auto data = std::format("Data #{}", ++m_numberOfDataBlocksWritten);
+
+        // Write data to the client.
+        m_data.set_data(data);
+        m_data.set_sequencenumber(m_numberOfDataBlocksWritten);
+        *m_data.mutable_status() = m_writeStatus;
+        StartWrite(&m_data);
+    } else {
+        // No more data to write.
+        Finish(::grpc::Status::OK);
+    }
+}
+
+void
+DataStreamWriter::HandleFailure(const std::string errorMessage)
```

Comment:
This is accepting a copy of the string, and then I believe another copy is made when passed to `set_message`. It probably makes sense to accept a `const std::string&` here instead, which will avoid 1 of those copies.

Also, in general, when accepting a copy/value of an argument, it's usually a good idea to accept it as non-`const` so the function can use it however it likes, since it owns that memory.